### ISSUE TITLE
regression: TypeAhead select now has a browser-based pop-up

### ISSUE THREAD

**Issue by garrett** [patternfly/patternfly-react#6741]
In Cockpit, the TypeAhead select now has a browser-based pop-up on top of the PatternFly typeahead. (This also affects every project using the TypeAhead of PatternFly React. It's not Cockpit-specific.)

This is due to the PR https://github.com/patternfly/patternfly-react/pull/6623 landing which actually _causes_ the bug that https://github.com/patternfly/patternfly-react/issues/6625 talks about.

There is no "never" value for the autocomplete attribute. The correct value is "off" in almost every case. The only other value that makes sense is "new-password" for when an using typeahead for user names, else a password manager (including the built-in browser-based password managers) may show a pop-up.

Setting it to "never" means that it will autocomplete anything with a "never" attribute. This usually will result in no secondary popup on the first time, but a secondary popup is guaranteed on additional uses.

**Comment by garrett** [comment]
I left some details @ https://github.com/patternfly/patternfly-react/pull/6623#issuecomment-1005586144; quoting here:

> This change specifically makes it so we _do_ experience browser autocomplete on top of PatternFly autocomplete. In other words, this change is a breaking one that actually does the opposite of what is intended.
> 
> There's no "never" attribute listed on any documentation or spec pages, but there is an "off", which is suggest to be used in cases where autocomplete is not wanted. (The exception is user/pass forms where password managers still work with "off", but this isn't for autocomplete in most cases. In that specific case, [it's suggested to use "new-password"](https://developer.mozilla.org/en-US/docs/Web/Security/Securing_your_site/Turning_off_form_autocompletion#the_autocomplete_attribute_and_login_fields)... and this wouldn't be case for the typeahead select, _unless_ it's used for user selection. And it usually isn't used for user selection.)
> 
> - https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete
>   > "off"
>   >
>   > The browser is not permitted to automatically enter or select a value for this field. It is possible that the document or application provides its own autocomplete feature, or that security concerns require that the field's value not be automatically entered.
> - https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#autofill
>   > When wearing the autofill expectation mantle, the autocomplete attribute, if specified, must have a value that is an ordered set of space-separated tokens consisting of either a single token that is an ASCII case-insensitive match for the string "off", or a single token that is an ASCII case-insensitive match for the string "on", or autofill detail tokens.

**Comment by garrett** [comment]
Suggestion:

1. Revert #6623 to fix the general case.
2. Make a new PR that sets autocompletes to "new-user" when the input is set to "user", to work around password manager popups.

**Comment by jelly** [comment]
The issue can be seen here with the latest cockpit-podman.

![image](https://user-images.githubusercontent.com/67428/148211992-c2ae7c72-86ab-4948-a647-dd8ad677c178.png)

**Comment by garrett** [comment]
After digging around a little more, it seems that Chrome is actively breaking the web and neither of the standards work: https://stackoverflow.com/questions/48150635/autocomplete-new-password-ignored-by-chrome-63-in-windows/49204211#49204211

But if you set it to another value other than the ones that are supposed to work, you break all non-Chrome browsers, as they will autocomplete based on that value.

I guess setting some string + unique ID (across sessions) for Chrome is the way to work around this, and do the right thing for other browsers that are actually following the spec instead of changing their behavior every so often (like Chrome).

Here's an example snippet that would do the right thing, based on browser:

```js
// Set autocompleteValue to off, unless the browser is Chrome (which isn't following autocomplete standards).
// Chrome gets a unique work-around value to prevent matching the same value more than once. (It would autocomplete to nothing as a workaround.)
var autocompleteValue = navigator.userAgent.match("Chrome") ? "pf--do-not-autocomplete-" + crypto.randomUUID() : "off";
```

It's a hack and I detest browser sniffing, but it's the cleanest way I could think of to do the right thing for all browsers except those who announce themselves as Chrome. (All Chromium-based browsers also announce themselves as Chrome, but Firefox and WebKit-based browsers do not.)

**Comment by tlabaj** [comment]
closed by #6846
